xknx/core/payload_reader.py

```diff
@@ -0,0 +1,109 @@
+"""
+Module for sending and receiving arbitrary payloads from the KNX bus.
+
+The module will
+* ... send the payload to the selected address.
+* ... register a callback for receiving telegrams within telegram queue.
+* ... check if received telegrams have the correct type address.
+* ... store the received telegram for further processing.
+"""
+import asyncio
+import logging
+from typing import TYPE_CHECKING, Optional, Type, Union
+
+from xknx.telegram import GroupAddress, IndividualAddress, Telegram
+from xknx.telegram.apci import APCI
+
+if TYPE_CHECKING:
+    from xknx.xknx import XKNX
+
+logger = logging.getLogger("xknx.log")
+
+
+class PayloadReader:
+    """Class for sending a request and waiting for a response on the KNX bus."""
+
+    # pylint: disable=too-many-instance-attributes
+
+    def __init__(
+        self,
+        xknx: "XKNX",
+        address: Union[GroupAddress, IndividualAddress],
+        timeout_in_seconds: int = 1,
+    ) -> None:
+        """Initialize PayloadReader class."""
+        self.xknx = xknx
+        self.address = address
+        self.timeout_in_seconds = timeout_in_seconds
+
+        self.response_received_or_timeout = asyncio.Event()
+        self.success = False
+        self.timeout_handle: Optional[asyncio.TimerHandle] = None
+        self.received_telegram: Optional[Telegram] = None
+
+    def reset(self) -> None:
+        """Reset reader for next send."""
+        self.response_received_or_timeout = asyncio.Event()
+        self.success = False
+        self.timeout_handle = None
+        self.received_telegram = None
+
+    async def send(
+        self, payload: APCI, response_class: Optional[Type[APCI]] = None
+    ) -> Optional[APCI]:
+        """Send group read and wait for response."""
+        self.reset()
+
+        cb_obj = self.xknx.telegram_queue.register_telegram_received_cb(
+            lambda t: self.telegram_received(t, response_class)
+        )
+
+        await self.xknx.telegrams.put(
+            Telegram(destination_address=self.address, payload=payload)
+        )
+        await self.start_timeout()
```

Comment:
Have you had a look at #523 https://github.com/XKNX/xknx/pull/523#discussion_r537895689
I've changed how these timeouts work in the rest of xknx. 
This way it can get canceled without leaving useless callbacks and I think it's easier readable.

Reply:
Fixed that in 94f1300.